src/core/ir/discriminator.rs

```diff
@@ -0,0 +1,1121 @@
+
+                    seen_required_in.insert(*field_info.required_in);
+
+                    !drop
+                })
+                .collect()
+        };
+
+        let discriminator = Self { fields_info, types };
+
+        tracing::debug!(
+            "Generated discriminator for union type '{union_name}':\n{discriminator:?}",
+        );
+
+        Valid::succeed(discriminator)
+    }
+
+    pub fn resolve_type(&self, value: &Value) -> Result<TypeName> {
+        if let Value::List(list) = value {
+            let results: Result<Vec<_>> = list
+                .iter()
+                .map(|item| Ok(self.resolve_type_for_single(item)?.to_string()))
+                .collect();
+
+            Ok(TypeName::Vec(results?))
+        } else {
+            Ok(TypeName::Single(
+                self.resolve_type_for_single(value)?.to_string(),
+            ))
+        }
+    }
+
+    fn resolve_type_for_single(&self, value: &Value) -> Result<&str> {
+        let Value::Object(obj) = value else {
+            bail!("Value expected to be object");
+        };
+
+        let mut possible_types = BitVec::set_after(self.types.len());
+
+        for (field, info) in &self.fields_info {
+            if obj.contains_key(field.as_str()) {
+                *possible_types &= *info.presented_in;
+            } else {
+                *possible_types &= !*info.required_in;
+            }
+
+            if possible_types.is_empty() {
+                // no possible types, something is wrong with the resolved value
+                bail!("Failed to find corresponding type for value")
+            }
+
+            if !possible_types.is_covering_multiple_types() {
+                // we've got only one possible type so return it
+                // even despite the fact the value could be totally wrong if we check other
+                // fields. But we want to cover positive cases and we want to do
+                // it as soon as possible and the wrong value will be probably
+                // wrong anyway to use later
+                return Ok(possible_types.first_covered_type(&self.types));
+            }
+        }
+
+        // we have multiple possible types. Return the first one
+        // that is defined earlier in config
+        Ok(possible_types.first_covered_type(&self.types))
+    }
+}
+
+/// BitVecesentation for set of types if some condition is hold.
+/// The condition is BitVecesented as a bit inside the usize number
+/// where bit position from the right in binary BitVecesentation of usize
+/// is the index of type in the set and the if the value of the bit is
+/// 1 then the condition is hold
+#[derive(Copy, Clone, Default)]
+struct BitVec(usize);
+
+impl std::fmt::Debug for BitVec {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("{:0b}", self.0))
+    }
+}
+
+impl Deref for BitVec {
```

Comment:
Not a good idea to implement Deref here. We should use Deref when the primary value is what's inside for eg: Box, Arc, Mutex etc. BitVec, is a great DS that encapsulates the idea of a vector for boolean values. We could have a `as_usize` method if really needed.

Reply:
replaced Deref/DerefMut with required operations traits

Reply:
Rust docs: https://doc.rust-lang.org/1.23.0/std/ops/trait.Deref.html
> Deref should only be implemented for smart pointers to avoid confusion.

Also look at — https://stackoverflow.com/questions/45086595/is-it-considered-a-bad-practice-to-implement-deref-for-newtypes

Reply:
well, there are mixed opinions on it based on discussion =)

But, I agree that we shouldn't overuse it and I've already replaced it